skimage/morphology/_rolling_ball_cy.pyx

```diff
@@ -0,0 +1,130 @@
+    offsets : (N) ndarray
+        Array of positions indicating which windows the convolution should be
+        applied to
+
+    Returns
+    -------
+    out_data : (N) ndarray
+        1D Array holding the result of the kernel being applied to the windows
+
+    See Also
+    --------
+    rolling_ellipsoid
+    """
+    
+    cdef DTYPE_FLOAT[::1] out_data = np.zeros(offsets.size, dtype=windows.base.dtype)
+    cdef Py_ssize_t offset, offset_idx, offsets_size, kern_x, kern_y
+    cdef DTYPE_FLOAT min_value, tmp
+    offsets_size = offsets.size
+
+    for offset_idx in prange(offsets_size, nogil=True):
+        offset = offsets[offset_idx]
+        min_value = INFINITY
+        for kern_y in range(kernel.shape[0]):
+            for kern_x in range(kernel.shape[1]):
+                tmp = (windows[offset, kern_y, kern_x] +
+                       cap_height[kern_y, kern_x]) * kernel[kern_y, kern_x]
+                if not min_value <= tmp:
+                    min_value = tmp
+                    if isnan(tmp):
+                        break
+            if isnan(min_value):
+                break
+        out_data[offset_idx] = min_value
+
+    return out_data.base
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def apply_kernel(DTYPE_FLOAT[:,:,:] windows,
+                 DTYPE_FLOAT[:, ::1] kernel,
+                 DTYPE_FLOAT[:, ::1] cap_height,
+                 Py_ssize_t[::1] offsets):
+    """
+    apply_kernel_flat(windows, kernel, cap_height)
+
+    Apply a custom kernel to a windowed view of an image.
+
+    This function is the critical piece of code for 
+    `morphology.rolling_ellipsoid`. It was moved to cython for speed.
+
+    Parameters
+    ----------
+    windows : (N, K1, K2) ndarray
+        A windowed view into an nD image. K1 and K2 are the dimensions of the
+        window and N is the number of windows into the image.
+    kernel : (K1, K2) ndarray
+        Indicates if pixel inside the window belongs to the kernel. 
+        `kernel[x,y] == 1` if the pixel is inside, ``kernel[x,y] == np.Inf`` 
+        otherwise.
+    cap_height : (K1, K2) ndarray
+        Indicates the height/intensity of the ellipsoid at position ``(x,y)``
+    offsets : (N) ndarray
+        1D Array of positions indicating which windows the convolution should
+        be applied to
+
+    Returns
+    -------
+    out_data : (N) ndarray
+        1D Array holding the result of the kernel being applied to the windows
+
+    See Also
+    --------
+    rolling_ellipsoid
+
+    Notes
+    -----
+    - This function assumes that the image doesn't contain ``NaN``s; this 
+    assumption allows for faster code (better compiler optimization).
+    """
+    
+    cdef DTYPE_FLOAT[::1] out_data = np.zeros(offsets.size, dtype=windows.base.dtype)
+    cdef Py_ssize_t offset, offset_idx, offsets_size, kern_x, kern_y
+    cdef DTYPE_FLOAT min_value, tmp
+    offsets_size = offsets.size
+
+    for offset_idx in prange(offsets_size, nogil=True):
```

Comment:
Would it be possible to add a `num_threads` value here (which could be None), and a `num_workers` value to `apply_kernel` and the Python functions? Therefore users would be able to choose (and avoid over-subscription if they have already using several threads or processes). See for example https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/_cycle_spin.py#L49

Reply:
I agree that a `num_threads` variable would be useful. It's just not so easy to get it to work in the same way as the example with OpenMP; at least I am a bit stuck on it.

The problem case is `num_threads=None` (default). Intuitively you might assume `num_threads=multiprocessing.cpu_count()` (max threads), but I guess it should rather default to `openmp`'s `OMP_NUM_THREADS`. I want to avoid checking environment variables directly - that seems very hacky - but didn't find a way to use openMP to give me that number yet.

Cython's default for `num_threads` is also `None`, so I assume they take care of this under the hood, and we should leverage that. The problem is that I can't seem to be able to figure out how I can have `num_threads` be either `None` or `int` without using a python object, which is not allowed inside `nogil`, and `nogil` is required for parallel processing.

I checked the places where the library already uses `openmp`. There is only one (_cascade.pyx) and it doesn't allow to specify the number of threads. On this note, it seems a bit strange to rely on dask AND openMP to achieve multithreadding. What is the plan for this for the future?

Reply:
I managed to solve this. `openmp.omp_get_max_threads()` actually provides the number of threads OpenMP should use, and not the maximum number of cores; somehow I confused that.